plugin-feature-doc.md

```diff
@@ -0,0 +1,111 @@
+# Plugin API is simple and well-documented
+
+The current plugin API has an overly large surface area, poor documentation,
+and unclear names.  The objective is to improve this, focusing on the UX of
+primary use-case: creating new service authenticators.
+
+Aha Card: https://cyberark.aha.io/features/AAM-92
+
+- [Objective](#objective)
+- [Experience (Summary)](#experience-summary)
+- [Experience (Detailed)](#experience-detailed)
+- [Technical Details](#technical-details)
+- [Testing](#testing)
+- [Documentation](#documentation)
+- [Open Questions](#open-questions)
+- [Stories](#stories)
+- [Future Work](#future-work)
+
+### Objective
+
+- Simplify Plugin API so authors of new service authenticators have a good
+  experience which requires learning only concepts relevant to their goal.
+- Write clear documentation to make the simplified API accessible to new
+  contributors
+
+### Experience (Summary)
+
+The new plugin API will support writing new service authenticators only.  Only
+a minimal interface will be required.
+
+### Experience (Detailed)
+
+Authors will need to implement only `PluginInfo` and a new
+`ServiceAuthenticator` interface.
+
+`APIVersion` will be moved into `PluginInfo`, and the `plugin_type` will be
+explicitly specified:
+
+```go
+var PluginInfo = map[string]string{
+  "api_version": "2.0",
+  "plugin_types": "service_authenticator",
+  "version":     "0.0.7",
+  "id":          "test-plugin",
+  "name":        "Test Plugin",
+  "description": "Test plugin to demonstrate plugin functionality",
+}
+```
+
+Implementing authenticators will look like:
+
+```go
+func ServiceAuthenticators() map[string]func(plugin_v2.AuthenticatorOptions) plugin_v2.Authenticator {
+  return map[string]func(plugin_v2.AuthenticatorOptions) plugin_v2.Authenticator{
+    "sample-authenticator-plugin": samplePlugin.NewAuthenticator,
+  }
+}
+```
+
+With the precise interface of an `Authenticator` TBD.
+
+##### Assumptions
+
+- The primary interest in plugins, at least at first, will be in writing new service authenticators.
+
+##### Workflow
+
+1. Write and compile one or more authenticators according to the above
+1. Place the resulting `.so` files in the correct directory (currently `/usr/local/lib/secretless`)
+1. Restart secretless
+
+### Technical Details
+
+- Per above, we'll refactor so that only the new, above described minimal API is public
+- All the old interfaces will be moved into internal
+- We'll have a generic type doing what's currently repeated by all listeners
+- We'll need to deal with the exceptions represented by the http handler --
+  what we were calling "sub-handlers" in previous discussions (as a placeholder
+  name).
+- Once handlers and listeners are combined, go back and update our config code
+  to use them directly.
+
```

Comment:
I don't have a clear picture from this of what's happening in the internal code. we're going to change the interface - move some parts of it to `internal/`, revise the public pieces that remain. but:

- are we also updating the internal code to implement the v2 interface?
- in a related question, are we going to reorganize the internal code to remove the distinction between listeners and handlers? eg put code in a single (maybe) `authenticators` directory, remove internal references to listeners/handlers, etc?
- is the v2 config going to become part of the public v2 plugin interface, or is it going to remain its own package?
- are we going to update the internal components that still use the v1 config, or will that tech debt remain after this effort?
- what other tech debt are we not capturing in what is described above, especially around the fundamental concepts that are part of the code? it would be good to capture that in more detail in the `future work` section

Reply:
> are we also updating the internal code to implement the v2 interface?

yes.  the current "adapter" approach we're using was intended to be temporary, to keep us focused on just delivering the config changes.

> in a related question, are we going to reorganize the internal code to remove the distinction between listeners and handlers? eg put code in a single (maybe) authenticators directory, remove internal references to listeners/handlers, etc?

yep, that's the basic plan

> is the v2 config going to become part of the public v2 plugin interface, or is it going to remain its own package?

i haven't thought super deeply about this, and atm i don't feel strongly one way or the other.

> are we going to update the internal components that still use the v1 config, or will that tech debt remain after this effort?

i think we should completely eliminate it completely from the code, and no longer support it.  if ending support for it isn't an option for some reason, we should still end support for it.  if it's *really* not an option, we can support it using an adapter approach: ie, parse it and convert it internally into a v2 config object.

> are we going to update the internal components that still use the v1 config, or will that tech debt remain after this effort?

update.

> what other tech debt are we not capturing in what is described above, especially around the fundamental concepts that are part of the code? it would be good to capture that in more detail in the future work section

there's debt around the plugin manager, the connection managers, as well as others.  i have notes and even some local branches where i've started work.  i can update the document to include that stuff?   but technically it's not part of the "plugin" refactor (again, at least in my mind).  i can think about which parts are "most adjacent" though...